src/ast/nodes/shared/FunctionNode.ts

```diff
@@ -85,8 +86,16 @@ export default class FunctionNode extends NodeBase {
 	}
 
 	reassignPath(path: ObjectPath) {
-		if (path.length === 1 && path[0] === 'prototype') {
-			this.isPrototypeReassigned = true;
+		if (path.length === 1) {
+			if (path[0] === 'prototype') {
+				this.isPrototypeReassigned = true;
+			} else if (path[0] === UNKNOWN_KEY) {
+				this.isPrototypeReassigned = true;
+
+				// A reassignment of [UNKNOWN_PATH] is considered equivalent to having lost track
+				// which means the return expression needs to be reassigned as well
+				this.scope.getReturnExpression().reassignPath(UNKNOWN_PATH);
+			}
```

Comment:
There are more situations where a reassignment happens. In all situations where the path is NOT UNKNOWN_PATH, there is an actual assignment taking place. More situations where we reassign UNKNOWN_PATH:
- there is more than one return expression
- a previously optimized conditional is deoptimized
- something is an export of the bundle
- something is an array element
- a variable has more than one declaration
- in any assignment, not only is the EMPTY_PATH of the left side reassigned, we also reassign UNKNOWN_PATH of the right side as we are no longer tracking mutations here (this one is REALLY important)
- this also goes for assignment patterns
- something is yielded (this one is new)

But you are right, this double meaning of reassignPath(UNKNOWN_PATH) should warrant a new name. I will go for `deoptimizePath` and probably rename the existing `deoptimize` methods to `deoptimizeCache` to better distinguish between the two.

Reply:
In cases where `reassignPath` really does mean `reassignPath` could we retain the meaning, or is that a big refactor? I just think it seems like there should be a semantic distinction between `reassignPath` which refers to reassigning a binding (which might be possible to track in future), and `deoptimizePath`?

Reply:
Nice changing to `deoptimizePath` though that seems much clearer to me!

Or do you think that `reassignPath` for actual reassignments is always about deoptimization, and that if we were to have better tracking here, we'd do it another way?

Reply:
Also `deoptimizeCache` sounds great, nice to be clear on what exactly is the deoptimization!

Reply:
The hope is that ultimately with SSA, all remaining reassignments will be "real" deoptimizations where an entity is out of sight of the algorithm. But also in its current form, deoptimize is probably a better name because we never reassign *to* anything but mark it as evil and unreliable ;)